src/hardware/adlib_gold.cpp

```diff
@@ -0,0 +1,352 @@
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "adlib_gold.h"
+#include "checks.h"
+
+CHECK_NARROWING();
+
+// Yamaha YM7128B Surround Processor emulation
+// -------------------------------------------
+
+SurroundProcessor::SurroundProcessor(const uint16_t sample_rate) : chip(nullptr)
+{
+	chip = (YM7128B_ChipIdeal *)malloc(sizeof(YM7128B_ChipIdeal));
+
+	YM7128B_ChipIdeal_Ctor(chip);
+	YM7128B_ChipIdeal_Setup(chip, sample_rate);
+	YM7128B_ChipIdeal_Reset(chip);
+	YM7128B_ChipIdeal_Start(chip);
+}
+
+SurroundProcessor::~SurroundProcessor()
+{
+	YM7128B_ChipIdeal_Stop(chip);
+	YM7128B_ChipIdeal_Dtor(chip);
+}
+
+union SurroundControlReg {
+	uint8_t data = 0;
+	bit_view<0, 1> din; // serial data
+	bit_view<1, 1> sci; // bit clock
+	bit_view<2, 1> a0;  // word clock
+};
+
+void SurroundProcessor::ControlWrite(const uint8_t val) noexcept
+{
+	SurroundControlReg reg = {val};
+
+	// Change register data at the falling edge of 'a0' word clock
+	if (control_state.a0 && !reg.a0) {
+		//		DEBUG_LOG_MSG("ADLIBGOLD: Surround: Write
+		// control register %d, data: %d",
+		// control_state.addr, control_state.data);
+
+		YM7128B_ChipIdeal_Write(chip, control_state.addr, control_state.data);
+	} else {
+		// Data is sent in serially through 'din' in MSB->LSB order,
+		// synchronised by the 'sci' bit clock. Data should be read on
+		// the rising edge of 'sci'.
+		if (!control_state.sci && reg.sci) {
+			// The 'a0' word clock determines the type of the data.
+			if (reg.a0)
+				// Data cycle
+				control_state.data = static_cast<uint8_t>(
+				                             control_state.data << 1) |
+				                     reg.din;
+			else
+				// Address cycle
+				control_state.addr = static_cast<uint8_t>(
+				                             control_state.addr << 1) |
+				                     reg.din;
+		}
+	}
+
+	control_state.sci = reg.sci;
+	control_state.a0  = reg.a0;
+}
+
+AudioFrame SurroundProcessor::Process(const AudioFrame &frame) noexcept
+{
+	YM7128B_ChipIdeal_Process_Data data = {};
+
+	data.inputs[0] = frame.left + frame.right;
+
+	YM7128B_ChipIdeal_Process(chip, &data);
+
+	return {data.outputs[0], data.outputs[1]};
+}
+
+// Philips Semiconductors TDA8425 hi-fi stereo audio processor emulation
+// ---------------------------------------------------------------------
+
+StereoProcessor::StereoProcessor(const uint16_t _sample_rate)
+        : sample_rate(_sample_rate)
+{
+	constexpr auto allpass_freq = 400.0;
+	constexpr auto q_factor     = 1.7;
+	allpass.setup(sample_rate, allpass_freq, q_factor);
```

Comment:
Maybe a similar assert for positive `sample_rate`?
(I'm not sure if a zero-value would throw the TDA8425 code off).

Reply:
It doesn't hurt to add it.